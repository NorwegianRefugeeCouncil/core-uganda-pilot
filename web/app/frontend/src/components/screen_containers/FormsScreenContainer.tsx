import { FormDefinition } from "core-js-api-client/lib/types/types";
import React from "react";

import { FormsScreenContainerProps } from "../../types/screens";
import client from "../../utils/clients";
import { FormsScreen } from "../screens/FormsScreen";

export const FormsScreenContainer = ({
    navigation,
    route,
}: FormsScreenContainerProps) => {
    const [forms, setForms] = React.useState<FormDefinition[]>();
    const [isLoading, setIsLoading] = React.useState(true);
    React.useEffect(() => {
        client.listForms({}).then((data) => {
            setIsLoading(false);
            setForms(data.response?.items);
        });
    });
    return (
        <FormsScreen
            isLoading={isLoading}
            forms={forms}
            navigation={navigation}
        />
    );
};
