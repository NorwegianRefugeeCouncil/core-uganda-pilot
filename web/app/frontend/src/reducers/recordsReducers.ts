import {Reducer} from "react";
import _ from 'lodash';
import {FormDefinition, Record} from "core-js-api-client/lib/types/types";

export type RecordsStoreProps = {
    formsById: {
        [key: string]: {
            definition: FormDefinition,
            records: Record[],
            recordsById: { [key: string]: Record; }
            localRecords: string[]
        }
    }
}

type RecordsAction = {
    type: string, // TODO use enum type
    payload: any
}

export const initialRecordsState: RecordsStoreProps = {
    formsById: {}
};

export enum RECORD_ACTIONS {
    GET_RECORDS = 'GET_RECORDS',
    GET_LOCAL_RECORDS = 'GET_LOCAL_RECORDS',
    ADD_LOCAL_RECORD = 'ADD_LOCAL_RECORD'
}

export const recordsReducer: Reducer<RecordsStoreProps, RecordsAction> = (state: RecordsStoreProps, action: RecordsAction) => {
    const {formId} = action.payload

    switch (action.type) {
        case RECORD_ACTIONS.GET_RECORDS:
            const {records} = action.payload
            return {
                formsById: {
                    ...state.formsById,
                    [formId]: {
                        ...state.formsById[formId],
                        records: records,
                        recordsById: _.keyBy(records, 'id')
                    }
                }
            }
        case RECORD_ACTIONS.GET_LOCAL_RECORDS:
            const {localRecords} = action.payload
            return {
                formsById: {
                    ...state.formsById,
                    [formId]: {
                        ...state.formsById[formId],
                        localRecords
                    }
                }
            }
        case RECORD_ACTIONS.ADD_LOCAL_RECORD:
            const currentList = state.formsById[formId].localRecords || []
            const newList = currentList.concat([action.payload.localRecord])
            return {
                formsById: {
                    ...state.formsById,
                    [formId]: {
                        ...state.formsById[formId],
                        localRecords: newList
                    }
                }
            }
        default:
            throw new Error();
    }
}
