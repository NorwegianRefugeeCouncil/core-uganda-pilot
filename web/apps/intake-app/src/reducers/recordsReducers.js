"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.recordsReducer = exports.RECORD_ACTIONS = exports.initialRecordsState = void 0;
var lodash_1 = require("lodash");
exports.initialRecordsState = {
    formsById: {}
};
var RECORD_ACTIONS;
(function (RECORD_ACTIONS) {
    RECORD_ACTIONS["GET_RECORDS"] = "GET_RECORDS";
    RECORD_ACTIONS["GET_LOCAL_RECORDS"] = "GET_LOCAL_RECORDS";
    RECORD_ACTIONS["ADD_LOCAL_RECORD"] = "ADD_LOCAL_RECORD";
})(RECORD_ACTIONS = exports.RECORD_ACTIONS || (exports.RECORD_ACTIONS = {}));
var recordsReducer = function (state, action) {
    var _a, _b, _c;
    var formId = action.payload.formId;
    switch (action.type) {
        case RECORD_ACTIONS.GET_RECORDS:
            var records = action.payload.records;
            return {
                formsById: __assign(__assign({}, state.formsById), (_a = {}, _a[formId] = __assign(__assign({}, state.formsById[formId]), { records: records, recordsById: lodash_1["default"].keyBy(records, 'id') }), _a))
            };
        case RECORD_ACTIONS.GET_LOCAL_RECORDS:
            var localRecords = action.payload.localRecords;
            return {
                formsById: __assign(__assign({}, state.formsById), (_b = {}, _b[formId] = __assign(__assign({}, state.formsById[formId]), { localRecords: localRecords }), _b))
            };
        case RECORD_ACTIONS.ADD_LOCAL_RECORD:
            var currentList = state.formsById[formId].localRecords || [];
            var newList = currentList.concat([action.payload.localRecord]);
            return {
                formsById: __assign(__assign({}, state.formsById), (_c = {}, _c[formId] = __assign(__assign({}, state.formsById[formId]), { localRecords: newList }), _c))
            };
        default:
            throw new Error();
    }
};
exports.recordsReducer = recordsReducer;
