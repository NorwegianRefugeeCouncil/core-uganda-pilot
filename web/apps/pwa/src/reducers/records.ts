import {createAsyncThunk, createEntityAdapter, createSlice} from "@reduxjs/toolkit";
import {FieldDefinition, Record} from "../types/types";
import {RootState} from "../app/store";
import {RecordListRequest, RecordListResponse} from "@core/api-client";
import {
    formGlobalSelectors,
    selectFieldForSubForm,
    selectFormOrSubFormById,
    selectRootForm,
    selectSubFormForField
} from "./form";
import Client from "../app/client";

const adapter = createEntityAdapter<Record>({
    // Assume IDs are stored in a field other than `book.id`
    selectId: (db) => db.id,
    // Keep the "all IDs" array sorted based on book titles
    sortComparer: (a, b) => a.id.localeCompare(b.id),
})


export const fetchRecords = createAsyncThunk<RecordListResponse, RecordListRequest>(
    'records/fetch',
    async (request, thunkAPI) => {
        try {
            const response = await Client.listRecords(request)
            if (response.success) {
                return response
            } else {
                return thunkAPI.rejectWithValue(response)
            }
        } catch (err) {
            return thunkAPI.rejectWithValue(err)
        }
    }
)

export const recordsSlice = createSlice({
    name: 'records',
    initialState: {
        ...adapter.getInitialState(),
        fetchPending: false,
        fetchError: undefined as any,
        fetchSuccess: true
    },
    reducers: {
        addOne: adapter.addOne,
        addMany: adapter.addMany,
        removeAll: adapter.removeAll,
        removeMany: adapter.removeMany,
        removeOne: adapter.removeOne,
        updateMany: adapter.updateMany,
        updateOne: adapter.updateOne,
        upsertOne: adapter.upsertOne,
        upsertMany: adapter.upsertMany,
        setOne: adapter.setOne,
        setMany: adapter.setMany,
        setAll: adapter.setAll,
    },
    extraReducers: builder => {
        builder.addCase(fetchRecords.pending, (state, action) => {
            state.fetchSuccess = false
            state.fetchPending = true
        })
        builder.addCase(fetchRecords.rejected, (state, action) => {
            state.fetchSuccess = false
            state.fetchPending = false
            state.fetchError = action.payload
        })
        builder.addCase(fetchRecords.fulfilled, (state, action) => {
            state.fetchSuccess = true
            state.fetchPending = false
            state.fetchError = undefined
            if (action.payload.response?.items) {
                adapter.addMany(state, action.payload.response.items)
            }
        })
    }
})

export const recordActions = recordsSlice.actions
export const recordSelectors = adapter.getSelectors();
export const recordGlobalSelectors = adapter.getSelectors<RootState>(state => state.records)
export default recordsSlice.reducer

export const selectRecordsForForm = (state: RootState, formId: string, parentRecordId?: string) => {
    return recordGlobalSelectors.selectAll(state).filter(r => {
        return parentRecordId
            ? r.formId === formId && r.parentId === parentRecordId
            : r.formId === formId;
    })
}

export const selectSubFormCount = (recordId: string, fieldId: string): ((rootState: RootState) => number) => {
    return rootState => {
        const record = recordGlobalSelectors.selectById(rootState, recordId)
        if (!record) {
            return 0
        }
        const subForm = selectSubFormForField(fieldId)(rootState)
        if (!subForm) {
            return 0
        }
        const subFormId = subForm.id
        const allRecords = recordGlobalSelectors.selectAll(rootState)
        return allRecords.reduce((prec, current) => {
            if (current.parentId === recordId && current.formId === subFormId) {
                return prec++
            }
            return prec
        }, 0)
    }
}

export type SubRecordResult = {
    byFormId: { [formId: string]: Record[] }
    byFieldId: { [fieldId: string]: Record[] }
}

export const selectSubRecords: (state: RootState, recordId: string) => SubRecordResult = (state, recordId) => {

    const result: SubRecordResult = {
        byFieldId: {},
        byFormId: {},
    }

    const record = recordGlobalSelectors.selectById(state, recordId)
    if (!record) {
        return result
    }

    const form = selectFormOrSubFormById(state, record.formId)
    if (!form) {
        return result
    }

    const rootForm = selectRootForm(state, form.id)
    if (!rootForm) {
        return result
    }

    const fieldMap: { [formId: string]: FieldDefinition } = {}

    const subFormIds = new Set<string>()
    for (let field of form.fields) {
        if (field.fieldType.subForm) {
            let fieldForSubForm = selectFieldForSubForm(rootForm, field.fieldType.subForm.id);
            if (fieldForSubForm) {
                fieldMap[field.fieldType.subForm.id] = fieldForSubForm
            }
            subFormIds.add(field.fieldType.subForm.id)
        }
    }

    const allRecords = recordGlobalSelectors.selectAll(state)
    for (let candidateRecord of allRecords) {
        if (candidateRecord.parentId === recordId && subFormIds.has(candidateRecord.formId)) {
            const fieldForSubform = fieldMap[candidateRecord.formId]
            if (fieldForSubform) {
                if (!result.byFieldId.hasOwnProperty(fieldForSubform.id)) {
                    result.byFieldId[fieldForSubform.id] = []
                }
                result.byFieldId[fieldForSubform.id].push(candidateRecord)
                if (!result.byFormId.hasOwnProperty(candidateRecord.formId)) {
                    result.byFormId[candidateRecord.formId] = []
                }
                result.byFormId[candidateRecord.formId].push(candidateRecord)
            }
        }
    }

    return result

}


export const selectRecordsSubFormCounts: (formId?: string) => ((rootState: RootState) => { [recordId: string]: { [fieldId: string]: number } }) = formId => {
    return rootState => {

        if (!formId) {
            return {}
        }

        const result: { [recordId: string]: { [fieldId: string]: number } } = {}

        const form = formGlobalSelectors.selectById(rootState, formId)
        if (!form) {
            return {}
        }

        if (!form.fields){
            return {}
        }

        // maps which form ids correspond to which field ids [formId] -> [fieldId]
        const formIdFieldIdMap: { [formId: string]: string } = {}
        for (let formField of form?.fields) {
            if (!formField?.fieldType?.subForm) {
                continue
            }
            formIdFieldIdMap[formField.fieldType.subForm.id] = formField.id
        }

        // records for the given form
        const formRecords = selectRecordsForForm(rootState, formId)
        if (formRecords.length === 0) {
            return {}
        }

        // map from recordId -> record
        const formRecordsMap: { [key: string]: Record } = {}
        for (let record of formRecords) {
            formRecordsMap[record.id] = record
        }

        // all records
        const allRecords = recordGlobalSelectors.selectAll(rootState)

        for (let record of allRecords) {

            // record does not have a parent record, does not qualify as subform
            if (!record.parentId) {
                continue
            }
            // parent record is not part of the current form
            if (!formRecordsMap.hasOwnProperty(record.parentId)) {
                continue
            }
            // field is not part of the current form
            if (!formIdFieldIdMap.hasOwnProperty(record.formId)) {
                continue
            }
            // the field id for that sub record
            const recordFieldId = formIdFieldIdMap[record.formId]

            // construct result
            if (!result.hasOwnProperty(record.parentId)) {
                result[record.parentId] = {}
            }
            if (!result[record.parentId].hasOwnProperty(recordFieldId)) {
                result[record.parentId][recordFieldId] = 0
            }

            // increase
            result[record.parentId][recordFieldId]++
        }

        return result

    }
}

export function selectRecords(state: RootState, options: { formId?: string, parentId?: string }) {
    const allRecords = recordGlobalSelectors.selectAll(state)
    const result: Record[] = []
    for (let record of allRecords) {
        if (options.formId && record.formId !== options.formId) {
            continue
        }
        if (options.parentId && record.parentId !== options.parentId) {
            continue
        }
        result.push(record)
    }
    return result
}
