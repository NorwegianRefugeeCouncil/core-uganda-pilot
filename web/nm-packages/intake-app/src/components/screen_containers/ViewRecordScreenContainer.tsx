import React from 'react';
import {ViewRecordScreenContainerProps} from '../../types/screens';
import {FormDefinition} from 'core-js-api-client/lib/types/types';
import {useApiClient} from '../../utils/useApiClient';
import {useForm} from 'react-hook-form';
import {ViewRecordScreen} from '../screens/ViewRecordScreen';

export const ViewRecordScreenContainer = ({
                                              route,
                                              state,
                                          }: ViewRecordScreenContainerProps) => {
    const {formId, recordId} = route.params;

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();

    const client = useApiClient();
    const {control, reset} = useForm();

    React.useEffect(() => {
        client.getForm({id: formId}).then(data => {
            setForm(data.response);
        });
    }, [formId]);

    React.useEffect(() => {
        if (form) {
            reset(state.formsById[formId].recordsById[recordId].values);
            setIsLoading(false);
        }
    }, [form]);

    return <ViewRecordScreen isLoading={isLoading} control={control}/>;
};
