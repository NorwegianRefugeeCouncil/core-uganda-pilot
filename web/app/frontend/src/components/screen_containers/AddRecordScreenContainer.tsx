import { FormDefinition } from "core-js-api-client";
import { NetworkStateType } from "expo-network";
import React from "react";
import { useForm } from "react-hook-form";
import { Platform } from "react-native";
import { RECORD_ACTIONS } from "../../reducers/recordsReducers";
import { AddRecordScreenContainerProps } from "../../types/screens";
import client from "../../utils/clients";
import { getEncryptionKey } from "../../utils/getEncryptionKey";
import { getNetworkState } from "../../utils/getNetworkState";
import { getEncryptedLocalData, storeEncryptedLocalData } from "../../utils/storage";
import { AddRecordScreen } from "../screens/AddRecordScreen";

export const AddRecordScreenContainer = ({ route, dispatch }: AddRecordScreenContainerProps) => {
    const { formId, recordId } = route.params;

    const isWeb = Platform.OS === "web";

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();
    const [isConnected, setIsConnected] = React.useState(false);
    const [hasLocalData, setHasLocalData] = React.useState(false);

    const { control, handleSubmit, formState, reset } = useForm();

    React.useEffect(() => {
        async function fetches() {
            let form, localData, networkState;

            // react to network changes
            try {
                networkState = await getNetworkState();
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setIsConnected(networkState !== NetworkStateType.NONE);
            }

            //
            try {
                const data = await client().getForm({ id: formId });
                form = data?.response;
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setForm(form);
                setIsLoading(false);
            }

            // check for locally stored data on mobile device
            if (!isWeb && recordId) {
                try {
                    localData = await getEncryptedLocalData(recordId);
                } catch (error) {
                    console.error(error);
                } finally {
                    setHasLocalData(!!localData);
                    reset(localData);
                }
            }
        }
        fetches();
    });

    const onSubmitOffline = async (data: any) => {
        const key = getEncryptionKey();
        try {
            await storeEncryptedLocalData(recordId, key, data);
            dispatch({
                type: RECORD_ACTIONS.ADD_LOCAL_RECORD,
                payload: {
                    formId,
                    localRecord: recordId,
                },
            });
            setHasLocalData(true);
        } catch (e) {
            setHasLocalData(false);
        }
    };

    const onSubmit = (data: any) => {
        handleSubmit(async () => {
            if (isConnected || isWeb) {
                await client().createRecord({
                    object: { formId, values: data },
                });
            } else {
                await onSubmitOffline(data);
            }
        });
    };

    return (
        <AddRecordScreen
            form={form}
            control={control}
            onSubmit={onSubmit}
            formState={formState}
            isWeb={isWeb}
            hasLocalData={hasLocalData}
            isConnected={isConnected}
            isLoading={isLoading}
        />
    );
};
