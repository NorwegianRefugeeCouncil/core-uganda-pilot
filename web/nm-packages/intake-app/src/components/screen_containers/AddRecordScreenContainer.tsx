import { FormDefinition } from 'core-js-api-client/lib/types/types';
import { NetworkStateType } from 'expo-network';
import React from 'react';
import { useForm } from 'react-hook-form';
import { Platform } from 'react-native';

import { RECORD_ACTIONS } from '../../reducers/recordsReducers';
import { AddRecordScreenContainerProps } from '../../types/screens';
import { getEncryptionKey } from '../../utils/getEncryptionKey';
import { getNetworkState } from '../../utils/getNetworkState';
import {
    getEncryptedLocalData,
    storeEncryptedLocalData,
} from '../../utils/storage';
import { useApiClient } from '../../utils/useApiClient';
import { AddRecordScreen } from '../screens/AddRecordScreen';

export const AddRecordScreenContainer = ({
    route,
    dispatch,
}: AddRecordScreenContainerProps) => {
    const { formId, recordId } = route.params;

    const client = useApiClient();
    const isWeb = Platform.OS === 'web';

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();
    // const [simulateOffline, setSimulateOffline] = React.useState(!isWeb); // TODO: for testing, remove
    const [isConnected, setIsConnected] = React.useState(false);
    // const [showSnackbar, setShowSnackbar] = React.useState(!isConnected);
    const [hasLocalData, setHasLocalData] = React.useState(false);

    const { control, handleSubmit, formState, reset } = useForm();

    React.useEffect(() => {
        async function fetches() {
            let form, localData, networkState;

            // react to network changes
            try {
                networkState = await getNetworkState();
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setIsConnected(networkState === NetworkStateType.NONE);
            }

            //
            try {
                const data = await client.getForm({ id: formId });
                form = data?.response;
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setForm(form);
                setIsLoading(false);
            }

            // check for locally stored data on mobile device
            if (!isWeb && recordId) {
                try {
                    localData = await getEncryptedLocalData(recordId);
                } catch (error) {
                    console.error(error);
                } finally {
                    setHasLocalData(!!localData);
                    reset(localData);
                }
            }
        }
        fetches();
    });

    const onSubmitOffline = async (data: any) => {
        const key = getEncryptionKey();

        storeEncryptedLocalData(recordId, key, data)
            .then(() => {
                setHasLocalData(true);
                dispatch({
                    type: RECORD_ACTIONS.ADD_LOCAL_RECORD,
                    payload: {
                        formId,
                        localRecord: recordId,
                    },
                });
            })
            .catch(() => {
                setHasLocalData(false);
            });
    };

    const onSubmit = (data: any) => {
        handleSubmit(async () => {
            if (isConnected || isWeb) {
                await client.createRecord({ object: { formId, values: data } });
            } else {
                await onSubmitOffline(data);
            }
        });
    };

    return (
        <AddRecordScreen
            form={form}
            control={control}
            onSubmit={onSubmit}
            formState={formState}
            isWeb={isWeb}
            hasLocalData={hasLocalData}
            isConnected={isConnected}
            isLoading={isLoading}
        />
    );
};
