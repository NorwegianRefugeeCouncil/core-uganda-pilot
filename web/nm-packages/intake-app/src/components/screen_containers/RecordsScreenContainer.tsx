import React from 'react';
import { RecordsScreenContainerProps } from '../../types/screens';
import {RecordsScreen} from '../screens/RecordsScreen';
import { useApiClient } from '../../utils/useApiClient';
import { RECORD_ACTIONS } from '../../reducers/recordsReducers';

export const RecordsScreenContainer = ({
    navigation,
    route,
    state,
    dispatch,
}: RecordsScreenContainerProps) => {
    const client = useApiClient();
    const { formId, databaseId } = route.params;
    const [isLoading, setIsLoading] = React.useState(true);
    const { records, localRecords } = state.formsById[formId];

    React.useEffect(() => {
        client.listRecords({ formId, databaseId }).then(data => {
            dispatch({
                type: RECORD_ACTIONS.GET_RECORDS,
                payload: {
                    formId,
                    records: data.response?.items,
                },
            });
            setIsLoading(false);
        });
    }, []);
    return (
        <RecordsScreen
            navigation={navigation}
            isLoading={isLoading}
            records={records}
            localRecords={localRecords}
            formId={formId}
        />
    );
};
