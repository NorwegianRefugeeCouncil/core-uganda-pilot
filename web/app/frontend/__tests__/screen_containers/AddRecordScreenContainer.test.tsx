import React from "react";
import { NetworkStateType } from "expo-network";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import { AddRecordScreenContainer } from "../../src/components/screen_containers/AddRecordScreenContainer";
import { RECORD_ACTIONS } from "../../src/reducers/recordsReducers";

const mockGetNetworkStateCellular = jest.fn().mockResolvedValue(NetworkStateType.CELLULAR);
const mockGetNetworkStateOffline = jest.fn().mockResolvedValue(NetworkStateType.NONE);
jest.mock("../../src/utils/getNetworkState", () => {
  return {
    getNetworkState: () => mockGetNetworkStateCellular(),
  };
});

const mockHandleSubmit = jest.fn(fn => fn());
const mockReset = jest.fn();
const mockFormState = {};
const mockControl = {};

jest.mock("react-hook-form", () => {
  const hf = jest.requireActual("react-hook-form");
  return {
    ...hf,
    useForm: () => ({
      handleSubmit: (fn: any) => mockHandleSubmit(fn),
      formState: mockFormState,
      reset: (data: any) => mockReset(data),
      control: mockControl,
    }),
  };
});

interface Data {
  [key: string]: any;
}

const mockForm = {};
const mockGetForm = jest.fn().mockResolvedValue(mockForm);
const mockCreateRecord = jest.fn().mockResolvedValue(true);
jest.mock("../../src/utils/clients", () => {
  return () => ({
    getForm: (query: any) => mockGetForm(query),
    createRecorde: () => mockCreateRecord(),
  });
});

const mockEncryptionKey = "super snazzy secret key";
const mockGetEncryptionKey = jest.fn().mockReturnValue(mockEncryptionKey);
jest.mock("../../src/utils/getEncryptionKey", () => {
  return {
    getEncryptionKey: () => mockGetEncryptionKey(),
  };
});

const mockLocalData: Data = { arbitrary: "data" };
const mockGetEncryptedLocalData = jest.fn().mockResolvedValue(mockLocalData);
const mockStoreEncryptedLocalData = jest.fn().mockResolvedValue(true);
jest.mock("../../src/utils/storage", () => {
  return {
    getEncryptedLocalData: () => mockGetEncryptedLocalData(),
    storeEncryptedLocalData: (recordId: any, key: any, data: any) =>
      mockStoreEncryptedLocalData(recordId, key, data),
  };
});

const mockRoute = {
  params: { formId: "randomId", recordId: "randomId" },
};

const mockDispatch = jest.fn();

const mockProps = {
  route: mockRoute,
  dispatch: (action: any) => mockDispatch(action),
};

const Component = (props: any) => <AddRecordScreenContainer {...props} />;

describe(AddRecordScreenContainer.name, () => {
  test("renders correctly", async () => {
    const { toJSON } = await waitFor(async () => render(<Component {...mockProps} />));
    expect(toJSON()).toMatchSnapshot();
  });

  test("gets network state", async () => {
    await waitFor(async () => render(<Component {...mockProps} />));
    expect(mockGetNetworkStateCellular).toHaveBeenCalled();
  });

  test("fetches a form with the supplied formId", async () => {
    await waitFor(async () => render(<Component {...mockProps} />));
    expect(mockGetForm).toHaveBeenCalledWith({ id: mockRoute.params.formId });
  });

  test("calls useForm's reset method with local data", async () => {
    await waitFor(async () => render(<Component {...mockProps} />));
    expect(mockReset).toHaveBeenCalledWith(mockLocalData);
  });

  test("fetches locally stored date when on mobile", async () => {
    await waitFor(async () => render(<Component {...mockProps} />));
    expect(mockGetEncryptedLocalData).toHaveBeenCalled();
  });

  // FIX ME
  test.skip("stores local data on user submit when offline", async () => {
    jest.mock("../../src/utils/getNetworkState", () => ({
      getNetworState: () => mockGetNetworkStateOffline(),
    }));
    const { getByA11yLabel } = await waitFor(async () => render(<Component {...mockProps} />));

    fireEvent.press(getByA11yLabel("Submit"), mockLocalData);

    expect(mockHandleSubmit).toHaveBeenCalled();
    expect(mockGetEncryptionKey).toHaveBeenCalled();
    expect(mockStoreEncryptedLocalData).toHaveBeenCalledWith(
      mockRoute.params.recordId,
      mockEncryptionKey,
      mockLocalData
    );
    const expectedAddRecordAction = {
      type: RECORD_ACTIONS.ADD_LOCAL_RECORD,
      payload: {
        formId: mockRoute.params.formId,
        localRecord: mockRoute.params.recordId,
      },
    };
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(expectedAddRecordAction);
  });
});
