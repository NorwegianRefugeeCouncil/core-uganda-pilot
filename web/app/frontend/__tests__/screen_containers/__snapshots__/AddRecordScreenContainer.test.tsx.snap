// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecordScreenContainer renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Array [
      Object {
        "flex": 1,
        "flexGrow": 1,
      },
      Object {
        "alignItems": "flex-start",
        "flexGrow": 1,
        "justifyContent": "flex-start",
        "padding": 18,
      },
      Object {
        "backgroundColor": "#24303E",
        "color": "#000000",
      },
    ]
  }
>
  <View>
    <View>
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      >
        <Text>
          There is locally stored data for this individual.
        </Text>
      </View>
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <View
          accessibilityLabel="Submit"
          accessibilityRole="button"
          accessibilityState={Object {}}
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            SUBMIT
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
